forge/forge/agent/forge_agent.py

```diff
@@ -0,0 +1,224 @@
+        if they want the agent to continue or not.
+        """  # noqa: E501
+
+        step = await self.db.create_step(
+            task_id=task_id, input=step_request, is_last=False
+        )
+
+        proposal = await self.propose_action()
+
+        output = await self.execute(proposal)
+
+        if isinstance(output, ActionSuccessResult):
+            step.output = str(output.outputs)
+        elif isinstance(output, ActionErrorResult):
+            step.output = output.reason
+
+        return step
+
+    async def propose_action(self) -> ActionProposal:
+        self.reset_trace()
+
+        # Get directives
+        directives = self.state.directives.copy(deep=True)
+        directives.resources += await self.run_pipeline(DirectiveProvider.get_resources)
+        directives.constraints += await self.run_pipeline(
+            DirectiveProvider.get_constraints
+        )
+        directives.best_practices += await self.run_pipeline(
+            DirectiveProvider.get_best_practices
+        )
+
+        # Get commands
+        self.commands = await self.run_pipeline(CommandProvider.get_commands)
+
+        # Get messages
+        messages = await self.run_pipeline(MessageProvider.get_messages)
+
+        prompt: ChatPrompt = ChatPrompt(
+            messages=messages, functions=function_specs_from_commands(self.commands)
+        )
+
+        logger.debug(f"Executing prompt:\n{dump_prompt(prompt)}")
+
+        # Call the LLM and parse result
+        # THIS NEEDS TO BE REPLACED WITH YOUR LLM CALL/LOGIC
+        # Have a look at autogpt/agents/agent.py for an example (complete_and_parse)
+        proposal = ActionProposal(
+            thoughts="I cannot solve the task!",
+            use_tool=AssistantFunctionCall(
+                name="finish", arguments={"reason": "Unimplemented logic"}
+            ),
+        )
+
+        self.config.cycle_count += 1
+
+        return proposal
+
+    async def execute(self, proposal: Any, user_feedback: str = "") -> ActionResult:
+        tool = proposal.use_tool
+
+        # Get commands
+        self.commands = await self.run_pipeline(CommandProvider.get_commands)
+
+        # Execute the command
+        try:
+            for c in reversed(self.commands):
+                if tool.name in c.names:
+                    command = c
+
+            command_result = command(**tool.arguments)
+            if inspect.isawaitable(command_result):
+                command_result = await command_result
+
+            result = ActionSuccessResult(outputs=command_result)
+        except AgentTerminated:
+            result = ActionSuccessResult(outputs="Agent terminated or finished")
+        except AgentException as e:
+            result = ActionErrorResult.from_exception(e)
+            logger.warning(f"{tool} raised an error: {e}")
+
+        await self.run_pipeline(AfterExecute.after_execute, result)
+
+        logger.debug("\n".join(self.trace))
```

Comment:
Typecheck fails here: 
`forge_agent.py:198:30 - error: "command" is possibly unbound (reportPossiblyUnboundVariable)`

Reply:
fixed
```py
command: Optional[Command] = None
for c in reversed(self.commands):
    if tool.name in c.names:
        command = c

if command is None:
    raise AgentException(f"Command {tool.name} not found")
```